doc/mirror.pl.8.in
'\" t
.TH MIRROR.PL 8 "25 Dec 2005" "@PACKAGE_VERSION@" "@PACKAGE_NAME@ utils"

.SH NAME
mirror.pl \- mirror the files from the list on standard input

.SH SYNOPSIS
.B mirror.plsh
[
.IR OPTIONS
]

.SH DESCRIPTION
\fBmirror.pl\fR copies the files it gets from standard input to the directory
specified with the \-b option. This script \fIimplements\fR the backup policy. 
.PP
If a file already exists in the backup directory it will be moved and
will get a suffix. This suffix is of the form: '+DD.HH:MM'. This suffix
is thus accurate to the minute. Multiple changes to files within one
minute are never caught by \fBmirror.pl\fR.
.PP
`hist.sh' and `yesterday.sh' also work on files with this same suffix.

.SH OPTIONS
.TP
.B \-a
Write the uid and gid information to the file's extended user attributes:
\'r_uid' and 'r_gid'. This option only works if \fBattr\fR is available
on the system. And currently only for Linux.

This option is usefull when storing a remote backup. Normally these
backups are transferred by a non-root user. This user will not be able
to correctly set all uid/gid information. As a workaround
\fBmirror.pl\fR will set the extended user attributes to the uid/gid as
it got them from standard input.
.TP 
.B \-b backupdir
Alternate backup directory. Defaults to `/vol/backup/$HOSTNAME'.
.TP
.B \-c 
Used in conjunction with \fBrdup\fR's \-c. This will allow
\fBmirror.pl\fR to be used for remote backups.
.TP
.B \-r
Restore mode. Always write the file to the target directory and
don't rename existing files to a +DD.HH:MM suffix.
This mode also filters out any file with such a suffix - as you
probably don't want those in your restore.
.TP
.B \-v
Echo the files processed to standard error.
.TP
.B \-h
Short help message.
.TP
.B \-V
Show the version.

.SH SEE ALSO
hist.sh(1), yesterday.sh(1), rdup(8)


ChangeLog
   Oct 2006: 0.2.18 Miek <miek@miek.nl>
   -- don't know yet --
   * removed: hist.sh and yesterday.sh
     never used them myself, probably stopped working in 0.2.17


24 Sept 2006: 0.2.17 Miek <miek@miek.nl>
   -- non backwards compatible update + minor feature update --
   * rdup: fix compilation on Mac OS X (and prob. Solaris)
   * mirror.pl: remove the YYYYMM suffix for backup directory
   * mirror.sh: remove the YYYYMM suffix for backup directory
   * dump.sh will now add this YYYYMM suffix!
   * added rdup.cron example
   * added grep.pl, grep.pl.1 manual page
   * grep.pl/gzip.pl/crypt.pl: small refactoring to make them 
     smaller
   * sh-restore/ is REMOVED, there are now no special restore utils
   * Doc: moved all docs over to the perl versions of the utils.
   * Making backups with 0.2.17:
   	dump.sh -b /vol/backup/$HOSTNAME dir-to-back-up
	YYYYMM will be added

	mirror.pl will use the -b argument as-is, no magic
	there.
   * Restoring backups with 0.2.17:
   	resore.sh -b /tmp/RESTORE dir-to-restore

	and in /tmp/RESTORE you will find your restore


18 Sep 2006: 0.2.16 Miek <miek@miek.nl>
   -- bugfixes and moved to the perl utils --
   rdup:
   * some compilation updates for Solaris 10 (dirdf)
   * small documentation updates
   configure:
   * added check for GNU make (isn't used yet)
   scripts:
   * All sizes are report in MB
   * mirror.pl: major fixes: use the correct backup dir
     and correct -c.
     Allow multiple -b options to backup to different directories
     at the same time
     The -b argument is mandatory, but can now be given
     multiple times.
   * dump.sh: use the perl tools! Added -x option.
     The -b argument is mandatory, but can now be given
     multiple times.
     The $HOST directory isn't added by default! Now you need
     to specify: -b /vol/backup/$HOST
   * dump-here.sh: use perl tools too.
   * Added crypt.pl - a perl version crypt.sh


1 Sep 2006: 0.2.15 Miek <miek@miek.nl>
   scripts:
   * refactored mirror.sh, rorrim.sh, monthday.sh, mirror.pl
   * dump.sh: more quotes for dirs with spaces. Re-add trailing spaces 
     to path names
   * coding style changes to the perl scripts.
   * rdup: add -l (local) option. Disable size checking for
     files when doing a local backup and don't skip the file.
   * Documentation updates


1 Jul 2006: 0.2.14 Miek <miek@miek.nl>
   -- misc. bugfixes and new features --
   scripts:
   * use dd for "catching" the files' content
   * eliminate the use of 'head' as this didn't work on freebsd
   * dump-here.sh must be edited before use (added exit 1)
   * Added pl-tools/ directory for perl contributed scripts:
     added gzip.pl and mirror.pl both from Mark J Hewitt
     These are up to 50 times faster than the shell scripts!
   * Remove rb-tools dir from distribution - not uses & not finished
   * moved wrapper scripts (dump.sh, dump-here.sh and restore.sh) to
     sh-wrapper/ directory
   * implement good bash coding style(Remarks from Bob Proulx)
     use booleans in the scripts
     kill whitespace
   rdup:
   * Don't re-check directories for changed files sizes
     and only check for files that are to be added.
   * Compilation fix for gcc 2.95; don't use %zd in printf
   * check return value of ftruncate()


26 May 2006: 0.2.13 Miek <miek@miek.nl>
   -- BACKWARD INCOMPATIBLE CHANGES --
   * dump.sh: drop req. argument NAME. This is now an option: -N NAME
   -- bugfixes and improvements --
   * monthday.sh: rewritten again...this time hopefully for the last
     time. It now is simpeler and local_restore and remote_restore
     are almost identical.
   * rdup: when a file changes size during the printing, skip it (Mark J Hewitt)
   * rdup: when a filename has a newline in it, skip it (Mark J Hewitt)


10 May 2006: 0.2.12 Miek <miek@miek.nl>
  -- bugfixes --
   * dump.sh: fix a bug in detecting whether to do a
     full or an incremental dump (Bram de Kruijff)
   * mirror.sh, rorrim.sh: silence chown errors
   * mirror.sh, rorrim.sh: echo to >&2 
   * mirror.sh, dump.sh: completed extended attribute support
   * mirror.sh, rorrim.sh: when not root and a directory
     would get 000 permissions, change that to 700
   * rdup, msg() fixlet
   * doc: added quick-start tex document for SANE
   * doc: poster/ directory was added for SANE
   * doc: more updates (Bram de Kruijff)
   * monthday.sh: with +0 don't print files with a suffix (Bram de Kruijff)
     (added restore test for this)


05 May 2006: 0.2.11 Miek <miek@miek.nl>
  -- major feature enhancement + major bugfixes --
   * rdup is now feature complete
   * dump.sh, restore.sh: make remote backups really work
   * scripts: implemented sanity checks on the data received. 
   extended attributes: [linux only atm]
   * mirror.sh: added -a flag, only needed here.
   * rdup: -a flag: read extended attributes to restore uid/gid 
   * bandwidth shaping was added to dump.sh and restore.sh, depends on
     cstream
   * rorrim.sh, monthday.sh: cosmetic fixes
   * rdup: added a msg function to remove all fprintfs
   * documentation was updated and improved


01 May 2006: 0.2.10 Miek <miek@miek.nl>
  -- tiny bugfixes + major restore updates --
   * rdup: formalize file handling in rdup:
     - plain files are accepted on the cmd line (they
       already were, but generated a (harmless) warning)
     - if files/dirs do not exist rdup skips them, instead
       of quiting
   * rdup.8: documentation updates, and make it more clear
     that rdup handles file arguments just fine.
   * NEW: restore.sh: added this as a wrapper around rdup and
     monthday.sh and rorrim.sh
   * REMOVED: purge.sh, I want to focus on dump/restore. Everyone
     can write a rm -rf script
   * dump.sh, restore.sh: added -c user@remote switch to ease in 
     remote backups and remote restores.
   * man rdup.8: added -r/-m options.
   * doc: started tex documentation about rdup and its usage, 
     see the tex/ directory.

pl-tools/mirror.pl
#!/usr/bin/perl -w
#
# Copyright (c) 2005, 2006 Miek Gieben; Mark J Hewitt
# See LICENSE for the license
#
# This script implement a mirroring backup scheme
# -c is used for remote mirroring

use strict;

use Getopt::Std;
use File::Basename;
use File::Path;
use File::Copy;
use File::Copy::Recursive qw(dircopy rcopy fcopy);
use Fcntl qw(:mode);

my $ts = time;
my ($sec, $min, $hour, $mday, $mon, $year, $wday, $yday, $isdst) = localtime($ts);

my $S_MMASK = 07777;			# mask to get permission
my $S_ISDIR = 040000;
my $S_ISLNK = 0120000;

my $progName = basename $0;

my %opt;

getopts('acvhb:V', \%opt);

usage() if $opt{'h'};			# Usage for -h
version() if $opt{'V'};			# Version for -V
my $verbose = $opt{'v'};		# Verbose messages for -v
my $attr = $opt{'a'};			# Extended attributes set for -a
my $remote = $opt{'c'};			# Find content in pipe for -c
my $hostname = `hostname`;		# This hostname
chomp $hostname;
                                        # Location of backup root in local filesystem
my $backupDir = (defined $opt{'b'} ? $opt{'b'} : "/vol/backup/" . $hostname) . "/" . sprintf "%04d%02d", 1900+$year, 1+$mon;

my $attr_there = check_attr();		# Can we set extended attributes?

# Statistics
my $ftsize = 0;				# Total file size
my $ireg = 0;				# Number of regular files
my $idir = 0;				# Number of directories
my $ilnk = 0;				# Number of symbolic links
my $irm  = 0;				# Number of files removed

if ($remote)
{
    remote_mirror($backupDir);
}
else
{
    local_mirror($backupDir);
}

my $te = time;
printf STDERR "** #REG FILES  : %d\n", $ireg;
printf STDERR "** #DIRECTORIES: %d\n", $idir;
printf STDERR "** #LINKS      : %d\n", $ilnk;
printf STDERR "** #(RE)MOVED  : %d\n", $irm;
printf STDERR "** SIZE        : %d KB\n", $ftsize/1024;
printf STDERR "** STORED IN   : %s\n", $backupDir;
printf STDERR "** ELAPSED     : %d s\n", $te - $ts;

exit 0;


sub local_mirror
{
    my $dir = $_[0];

    while (($_ = <STDIN>))
    {
	chomp;
	my($mode, $uid, $gid, $psize, $fsize, $path) = split;
	my $dump = substr($mode, 0, 1);
	my $modebits = substr($mode, 1);

	sanity_check($dump, $modebits, $psize, $fsize, $uid, $gid);

	my $bits=$modebits & $S_MMASK; # permission bits
	my $typ = 0;
	$typ = 1 if ($modebits & $S_ISDIR) == $S_ISDIR;
	$typ = 2 if ($modebits & $S_ISLNK) == $S_ISLNK;
	print STDERR "$path\n" if $verbose;
	my $target = "$dir/$path";
	my $suffix = mirror_suffix($target);

	if ($dump eq '+')
	{				# add
	    if ($typ == 0)
	    {				# REG
		if ($suffix)
		{			# We only have a suffix if the file exists
		    rename $target, ($target . $suffix) or die "Cannot rename $target: $!";
		}
		copy($path, $target) or die "Copy failed: $!";
		chown $uid, $gid, $target;
		chown_attr($uid, $gid, $target);
		chmod $modebits, $target;
		$ftsize += $fsize;
		$ireg++;
	    }
	    elsif ($typ == 1)
	    {
		if (-f _ || -l _)
		{			# Type of entiry has changed
		    rename $target, ($target . $suffix) or die "Cannot rename $target: $!";
		}
		mkpath($target) unless -d _;
		chown $uid, $gid, $target;
		chown_attr($uid, $gid, $target);
		if ($> != 0 && $bits == 0)
		{
		    print STDERR "chmod $target to 700\n";
		    $bits = 0700;
		}
		chmod $bits, $target;
		$idir++;
	    }
	    elsif ($typ == 2)
	    {				# LNK; target id the content
		if ($suffix)
		{			# We only have a suffix if the file exists
		    rename $target, ($target . $suffix) or die "Cannot rename $target: $!";
		}
		copy $path, $target or die "$path: $!";
		chown $uid, $gid, $target;
		chown_attr($uid, $gid, $target);
		$ilnk++;
	    }
	}
	else
	{	     # move. It could be the stuff is not there, don't error on that.
	    if ($suffix)
	    {
		rename $target, ($target . $suffix) or die "Cannot rename $target: $!";
		$irm++;
	    }
	}
    }
}



sub remote_mirror
{
    my $dir = $_[0];

    while (($_ = <STDIN>))
    {
	chomp;

	my($mode, $uid, $gid, $psize, $fsize) = split;
	my $dump = substr($mode, 0, 1);
	my $modebits = substr($mode, 1);

	sanity_check($dump, $modebits, $psize, $fsize, $uid, $gid);
	my $path = "";
	read STDIN, $path, $psize;

	my $bits = $modebits & $S_MMASK;
	my $typ = 0;
	$typ = 1 if ($modebits & $S_ISDIR) == $S_ISDIR;
	$typ = 2 if ($modebits & $S_ISLNK) == $S_ISLNK;
	print STDERR "$path\n" if $verbose;
	my $target = "$dir/$path";
	my $suffix = mirror_suffix($target);
#
#       NOTE - mirror_suffix() calls stat() on the target.  We rely on thet below to avoid multiple expensive calls to stat.
#       If mirror_suffix is changed, or is naot called, it will be necassary to revise this assumtion.
#
	if ($dump eq '+')
	{				# add
	    if ($typ == 0)
	    {				# REG
		if ($suffix)
		{			# We only have a suffix if the file exists
		    rename $target, ($target . $suffix) or die "Cannot rename $target: $!";
		}
		open FILE, ">$target" or die "Cannot create $target: $!";
		if ($fsize != 0)
		{
		    copyout($fsize, *FILE);
		}
		chown $uid, $gid, *FILE;
		chown_attr($uid, $gid, $target);
		chmod $bits, *FILE;
		close FILE;
		$ftsize += $fsize;
		$ireg++;
	    }
	    elsif ($typ == 1)
	    {				# DIR
		if (-f _ || -l _)
		{			# Type of entiry has changed
		    rename $target, ($target . $suffix) or die "Cannot rename $target: $!";
		}
		mkpath($target) unless -d _;
		chown $uid, $gid, $target;
		chown_attr($uid, $gid, $target);
		if ($> != 0 && $bits == 0)
		{
		    print STDERR "chmod $target to 700\n";
		    $bits = 0700;
		}
		chmod $bits, $target;
		$idir++;
	    }
	    elsif ($typ == 2)
	    {				# LNK; target id the content
		if ($suffix)
		{			# We only have a suffix if the file exists
		    rename $target, ($target . $suffix) or die "Cannot rename $target: $!";
		}
		my $linkTarget = "";
		read STDIN, $linkTarget, $fsize;
		symlink $target, $linkTarget;
		chown $uid, $gid, $target;
		chown_attr($uid, $gid, $target);
		$ilnk++;
	    }
	}
	else
	{				# Remove
	    if ($suffix)
	    {			# We only have a suffix if the file exists
		rename $target, ($target . $suffix) or die "Cannot rename $target: $!";
	    }
	    $irm++;
	}
    }
}


sub chown_attr
{
    return unless $attr_there;
    return unless $attr;

    my $xuid = $_[0];
    my $xgid = $_[1];
    my $file = $_[2];

    if ($^O eq "linux")
    {
	system("attr -q -s r_uid -V$xuid \"$file\"");
	system("attr -q -s r_gid -V$xgid \"$file\"");
    }
}


# Eg:
# 2006-05-05 01:50:18.000000000 +0100
#         ^^ ^^^^^
# => "+05.01:50
#
sub mirror_suffix
{
    my $filename = $_[0];

    lstat($filename);
    return "" unless -e _;
    my ($sec, $min, $hour, $mday, $mon, $year, $wday, $yday, $isdst) = localtime((lstat(_))[9]);
    return sprintf "+%02d.%02d:%02d", $mday, $hour, $min;
}


sub sanity_check
{
    my $dump = $_[0];
    my $mode = $_[1];
    my $psize = $_[2];
    my $fsize = $_[3];
    my $uid = $_[4];
    my $gid = $_[5];

    die "$progName: dump ($dump) must be + or -"   if $dump ne "+" && $dump ne "-";
    die "$progName: mode ($mode) must be numeric"  unless $mode =~ "[0-9]+";
    die "$progName: psize ($psize) must be numeric" unless $psize =~ "[0-9]+";
    die "$progName: fsize ($fsize) must be numeric" unless $fsize =~ "[0-9]+";
    die "$progName: uid ($uid) must be numeric"   unless $uid =~ "[0-9]+";
    die "$progName: gid ($gid) must be numeric"   unless $gid =~ "[0-9]+";
}


sub check_attr
{
    if ($^O eq "linux")
    {
	map {return 1 if -x $_ . '/' . "attr"}  split(/:/, $ENV{'PATH'});
    }
    else
    {
	warn "Cannot set extended attributes";
    }
    return 0;
}


sub usage
{
    print "$progName [OPTIONS]\n\n";
    print "Mirror the files from the filelist of rdup\n\n";
    print "OPTIONS\n";
    print " -a      write extended attributes r_uid/r_gid with uid/gid\n";
    print " -c      process the file content also (rdup -c), for remote backups\n";
    print " -b DIR  use DIR as the backup directory, YYYYMM will be added\n";
    print " -v      print the files processed to stderr\n";
    print " -h      this help\n";
    print " -V      print version\n";
}

sub version
{
    print "** $progName: 0.2.12 (rdup-utils)"
}

sub copyout
{
    my $count = $_[0];
    my $pipe = $_[1];

    my $buf;
    my $n;

    while ($count > 4096)
    {
	$n = read STDIN, $buf, 4096;
	syswrite $pipe, $buf, $n;
	$count -= 4096;
    }
    if ($count > 0)
    {
	$n = read STDIN, $buf, $count;
	syswrite $pipe, $buf, $n;
    }
}
